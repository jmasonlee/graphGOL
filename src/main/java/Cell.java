public class Cell implements Comparable {
    public Integer y;
    public Integer x;

    public Cell(Integer x, Integer y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public int compareTo(Object o) {
        return 0;
    }
}
