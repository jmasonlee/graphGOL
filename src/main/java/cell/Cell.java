package cell;

public class Cell implements Comparable {
  public Integer y;
  public Integer x;

  public Cell(Integer x, Integer y) {
    this.x = x;
    this.y = y;
  }

  @Override
  public int compareTo(Object o) {
    return 0;
  }

  @Override
  public boolean equals(Object o) {

    if (o == this) {
      return true;
    }

    if (!(o instanceof Cell)) {
      return false;
    }

    Cell otherCell = (Cell) o;

    return otherCell.x == this.x && otherCell.y == this.y;
  }

  @Override
  public String toString(){
      return "[" + x + "," + y + "]";
  }
}
