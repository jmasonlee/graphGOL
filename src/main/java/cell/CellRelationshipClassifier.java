package cell;

public class CellRelationshipClassifier {
    public static Relationships classify(Cell cell, Cell cell1) {
        return Relationships.SELF;
    }

    public enum Relationships {
        DISCONNECTED(-1), SELF(0), NEIGHBOUR(1), COPARENT(2);

        public final int maxDistance;

        private Relationships(int maxDistance){
            this.maxDistance = maxDistance;
        }
    }


}
