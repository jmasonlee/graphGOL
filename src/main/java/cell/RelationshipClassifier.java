package cell;

public class RelationshipClassifier {
    public static Relationships classify(Cell cell, Cell cell1) {
        return Relationships.SELF;
    }

    public enum Relationships {
        DISCONNECTED(-1), SELF(0), NEIGHBOUR(1), COPARENT(2);

        public final int maxDistance;

        private Relationships(int maxDistance){
            this.maxDistance = maxDistance;
        }
    }


}
