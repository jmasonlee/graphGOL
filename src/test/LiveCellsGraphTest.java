import gameOfLife.cell.Cell;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;

public class LiveCellsGraphTest {
  @Test
  public void testCanBeCreatedWithArrayOfCells() {}

  @Test
  public void testCanGetCellsWithSpecifiedNumberOfNeighbours() {
    List<Cell> cells = new ArrayList<>();
    cells.add(new Cell(0, 0));
    cells.add(new Cell(0, 1));

    LiveCellsGraph liveCells = new LiveCellsGraph(cells);
    assertEquals(cells, liveCells.getCellsWithNumberOfNeighbours(1));
  }

  // Approvals - how
}
