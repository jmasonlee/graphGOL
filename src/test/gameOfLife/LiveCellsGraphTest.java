package gameOfLife;

import gameOfLife.TestUtils.CellCoverage;
import gameOfLife.cell.Cell;
import org.apache.commons.math3.util.CombinatoricsUtils;
import org.approvaltests.Approvals;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

public class LiveCellsGraphTest {

  @Test
  public void testCanGetCellsWithNoNeighbours() {
    int desiredNumberOfNeighbours = 0;
    StringBuilder testOutput = buildGraphAndFilterOnNeighbourCount(desiredNumberOfNeighbours);

    Approvals.verify(testOutput);
  }

  @Test
  public void testCanGetCellsWithOneNeighbour() {
    int desiredNumberOfNeighbours = 1;
    StringBuilder testOutput = buildGraphAndFilterOnNeighbourCount(desiredNumberOfNeighbours);

    Approvals.verify(testOutput);
  }

  @Test
  public void testCanGetCellsWithTwoNeighbours() {
    int desiredNumberOfNeighbours = 2;

    StringBuilder testOutput = buildGraphAndFilterOnNeighbourCount(desiredNumberOfNeighbours);

    Approvals.verify(testOutput);
  }

  @Test
  public void testCanGetCellsWithThreeNeighbours() {
    int desiredNumberOfNeighbours = 3;
    StringBuilder testOutput = buildGraphAndFilterOnNeighbourCount(desiredNumberOfNeighbours);

    Approvals.verify(testOutput);
  }

  private StringBuilder buildGraphAndFilterOnNeighbourCount(int desiredNumberOfNeighbours) {
    Cell centre = new Cell(0, 0);
    List<Cell> allNeighbours = generateAllPossibleNeighboursForCell(centre);

    List<List<Cell>> neighbourCombinations =
        allCombinationsOfNeighbours(allNeighbours, desiredNumberOfNeighbours);

    return filterOnNeighbourCountForAllNeighbourCombinations(
        centre, desiredNumberOfNeighbours, neighbourCombinations);
  }

  private List<Cell> generateAllPossibleNeighboursForCell(Cell cell) {
    Integer[] neighbourRange = new Integer[] {-1, 0, 1};

    Cell lowerLeftCell = new Cell(-1, -1);
    Cell upperRightCell = new Cell(1,1);
    List<Cell> allNeighbours = CellCoverage.generateAllPossibleCellsBetweenTwoCells(lowerLeftCell,upperRightCell);
    allNeighbours.remove(cell);

    return allNeighbours;
  }

  private Cell createNeighbour(Cell centre, int relativeX, int relativeY) {
    int neighbourX = centre.x + relativeX;
    int neighbourY = centre.y + relativeY;

    return new Cell(neighbourX, neighbourY);
  }

  private List<List<Cell>> allCombinationsOfNeighbours(
    List<Cell> allNeighbours, int desiredNumberOfNeighbours) {

    Iterator<int[]> neighbourIterator =
      createCombinationsIterator(allNeighbours, desiredNumberOfNeighbours);

    return iterateOverAllPossibleCombinations(allNeighbours, neighbourIterator);
  }

  private Iterator<int[]> createCombinationsIterator(
    List<Cell> allNeighbours, int desiredNumberOfNeighbours) {
    return CombinatoricsUtils.combinationsIterator(allNeighbours.size(), desiredNumberOfNeighbours);
  }

  private List<List<Cell>> iterateOverAllPossibleCombinations(
    List<Cell> allNeighbours, Iterator<int[]> neighbourIterator) {
    List<List<Cell>> neighbourCombinations = new ArrayList<>();

    while (neighbourIterator.hasNext()) {
      int[] neighbourIndices = neighbourIterator.next();

      List<Cell> neighbours = matchIndicesToNeighbours(allNeighbours, neighbourIndices);

      neighbourCombinations.add(neighbours);
    }

    return neighbourCombinations;
  }

  private List<Cell> matchIndicesToNeighbours(List<Cell> allNeighbours, int[] neighbourIndices) {
    List<Cell> neighbours = new ArrayList<>();

    for (int index : neighbourIndices) {
      neighbours.add(allNeighbours.get(index));
    }

    return neighbours;
  }

  private StringBuilder filterOnNeighbourCountForAllNeighbourCombinations(
      Cell centre, int desiredNumberOfNeighbours, List<List<Cell>> neighbourCombinations) {
    StringBuilder testOutput = new StringBuilder();

    neighbourCombinations.forEach(
        neighbourList -> testOutput.append(
            filterCellsOnNeighbourCount(
                centre, neighbourList, desiredNumberOfNeighbours)));

    return testOutput;
  }

  private String filterCellsOnNeighbourCount(
      Cell centre, List<Cell> neighbours, int desiredNumberOfNeighbours) {

    LiveCellsGraph graph = new LiveCellsGraph(getCellsUnderTest(centre, neighbours));
    List<Cell> originalCells = graph.getCells();
    List<Cell> filteredCells = graph.filterCellsByNeighbourCount(desiredNumberOfNeighbours);

    StringBuilder testOutput =
        formatFilteringResults(originalCells, filteredCells);

    return testOutput.toString();
  }

  private StringBuilder formatFilteringResults(
    List<Cell> originalCells, List<Cell> filteredCells) {
    StringBuilder testOutput = new StringBuilder();

    testOutput.append(originalCells);
    testOutput.append(" => ");
    testOutput.append(filteredCells);
    testOutput.append("\n");

    return testOutput;
  }

  private List<Cell> getCellsUnderTest(Cell centre, List<Cell> neighbours) {
    List<Cell> cells = new ArrayList<>();
    cells.add(centre);
    cells.addAll(neighbours);
    return cells;
  }
}
