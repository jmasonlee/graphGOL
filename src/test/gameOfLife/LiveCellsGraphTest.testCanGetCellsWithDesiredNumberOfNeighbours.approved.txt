[0] => 
 Select cells with 0 neighbours
	[[0,0]] => [[0,0]]
 
[1] => 
 Select cells with 1 neighbours
	[[0,0], [-1,-1]] => [[0,0], [-1,-1]]
	[[0,0], [-1,0]] => [[0,0], [-1,0]]
	[[0,0], [-1,1]] => [[0,0], [-1,1]]
	[[0,0], [0,-1]] => [[0,0], [0,-1]]
	[[0,0], [0,1]] => [[0,0], [0,1]]
	[[0,0], [1,-1]] => [[0,0], [1,-1]]
	[[0,0], [1,0]] => [[0,0], [1,0]]
	[[0,0], [1,1]] => [[0,0], [1,1]]
 
[2] => 
 Select cells with 2 neighbours
	[[0,0], [-1,-1], [-1,0]] => [[0,0], [-1,-1], [-1,0]]
	[[0,0], [-1,-1], [-1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1]] => [[0,0], [-1,0], [-1,1]]
	[[0,0], [-1,-1], [0,-1]] => [[0,0], [-1,-1], [0,-1]]
	[[0,0], [-1,0], [0,-1]] => [[0,0], [-1,0], [0,-1]]
	[[0,0], [-1,1], [0,-1]] => [[0,0]]
	[[0,0], [-1,-1], [0,1]] => [[0,0]]
	[[0,0], [-1,0], [0,1]] => [[0,0], [-1,0], [0,1]]
	[[0,0], [-1,1], [0,1]] => [[0,0], [-1,1], [0,1]]
	[[0,0], [0,-1], [0,1]] => [[0,0]]
	[[0,0], [-1,-1], [1,-1]] => [[0,0]]
	[[0,0], [-1,0], [1,-1]] => [[0,0]]
	[[0,0], [-1,1], [1,-1]] => [[0,0]]
	[[0,0], [0,-1], [1,-1]] => [[0,0], [0,-1], [1,-1]]
	[[0,0], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [1,0]] => [[0,0]]
	[[0,0], [-1,1], [1,0]] => [[0,0]]
	[[0,0], [0,-1], [1,0]] => [[0,0], [0,-1], [1,0]]
	[[0,0], [0,1], [1,0]] => [[0,0], [0,1], [1,0]]
	[[0,0], [1,-1], [1,0]] => [[0,0], [1,-1], [1,0]]
	[[0,0], [-1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [1,1]] => [[0,0]]
	[[0,0], [0,-1], [1,1]] => [[0,0]]
	[[0,0], [0,1], [1,1]] => [[0,0], [0,1], [1,1]]
	[[0,0], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [1,0], [1,1]] => [[0,0], [1,0], [1,1]]
 
[3] => 
 Select cells with 3 neighbours
	[[0,0], [-1,-1], [-1,0], [-1,1]] => [[0,0], [-1,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1]] => [[0,0], [-1,-1], [-1,0], [0,-1]]
	[[0,0], [-1,-1], [-1,1], [0,-1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1]] => [[0,0], [-1,0]]
	[[0,0], [-1,-1], [-1,0], [0,1]] => [[0,0], [-1,0]]
	[[0,0], [-1,-1], [-1,1], [0,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,1]] => [[0,0], [-1,0], [-1,1], [0,1]]
	[[0,0], [-1,-1], [0,-1], [0,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [0,1]] => [[0,0], [-1,0]]
	[[0,0], [-1,1], [0,-1], [0,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [1,-1]] => [[0,0], [0,-1]]
	[[0,0], [-1,0], [0,-1], [1,-1]] => [[0,0], [0,-1]]
	[[0,0], [-1,1], [0,-1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,0], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [0,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [1,0]] => [[0,0], [0,-1]]
	[[0,0], [-1,0], [0,-1], [1,0]] => [[0,0], [0,-1]]
	[[0,0], [-1,1], [0,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [0,1], [1,0]] => [[0,0], [0,1]]
	[[0,0], [-1,1], [0,1], [1,0]] => [[0,0], [0,1]]
	[[0,0], [0,-1], [0,1], [1,0]] => [[0,0], [1,0]]
	[[0,0], [-1,-1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [0,-1], [1,-1], [1,0]] => [[0,0], [0,-1], [1,-1], [1,0]]
	[[0,0], [0,1], [1,-1], [1,0]] => [[0,0], [1,0]]
	[[0,0], [-1,-1], [-1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,1], [1,1]] => [[0,0], [0,1]]
	[[0,0], [-1,1], [0,1], [1,1]] => [[0,0], [0,1]]
	[[0,0], [0,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [0,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [0,-1], [1,0], [1,1]] => [[0,0], [1,0]]
	[[0,0], [0,1], [1,0], [1,1]] => [[0,0], [0,1], [1,0], [1,1]]
	[[0,0], [1,-1], [1,0], [1,1]] => [[0,0], [1,0]]
 
[4] => 
 Select cells with 4 neighbours
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1]] => [[0,0], [-1,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,1]] => [[0,0], [-1,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [0,1]] => [[0,0], [-1,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [0,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [0,1]] => [[0,0], [-1,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [1,-1]] => [[0,0], [0,-1]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [1,-1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [1,0]] => [[0,0], [0,-1]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,1], [1,0]] => [[0,0], [0,1]]
	[[0,0], [-1,-1], [0,-1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [1,-1], [1,0]] => [[0,0], [0,-1]]
	[[0,0], [-1,0], [0,-1], [1,-1], [1,0]] => [[0,0], [0,-1]]
	[[0,0], [-1,1], [0,-1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [0,-1], [0,1], [1,-1], [1,0]] => [[0,0], [1,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,1], [1,1]] => [[0,0], [0,1]]
	[[0,0], [-1,-1], [0,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [0,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,1], [1,0], [1,1]] => [[0,0], [0,1]]
	[[0,0], [-1,1], [0,1], [1,0], [1,1]] => [[0,0], [0,1]]
	[[0,0], [0,-1], [0,1], [1,0], [1,1]] => [[0,0], [1,0]]
	[[0,0], [-1,-1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [0,-1], [1,-1], [1,0], [1,1]] => [[0,0], [1,0]]
	[[0,0], [0,1], [1,-1], [1,0], [1,1]] => [[0,0], [1,0]]
 
[5] => 
 Select cells with 5 neighbours
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [0,1]] => [[0,0], [-1,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [1,-1], [1,0]] => [[0,0], [0,-1]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,1], [1,0], [1,1]] => [[0,0], [0,1]]
	[[0,0], [-1,-1], [0,-1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [0,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0], [1,0]]
 
[6] => 
 Select cells with 6 neighbours
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [0,1], [1,-1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [0,1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [0,1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [0,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [0,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,1], [0,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
 
[7] => 
 Select cells with 7 neighbours
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [0,1], [1,-1], [1,0]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [0,1], [1,-1], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [0,1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,0], [0,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,-1], [-1,1], [0,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
	[[0,0], [-1,0], [-1,1], [0,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
 
[8] => 
 Select cells with 8 neighbours
	[[0,0], [-1,-1], [-1,0], [-1,1], [0,-1], [0,1], [1,-1], [1,0], [1,1]] => [[0,0]]
 
